### PR TITLE
Improve test coverage by mutation testing

### PR DESCRIPTION
**Pull request by TimWolla**
Using [mull](https://mull.readthedocs.io/en/0.26.0/GettingStarted.html):

```
cmake . -DURIPARSER_BUILD_DOCS=0 -DCMAKE_CXX_FLAGS="-O0 -fpass-plugin=/usr/lib/mull-ir-frontend-18 -g -grecord-command-line -fsanitize=address,undefined" -DCMAKE_C_FLAGS="-O0 -fpass-plugin=/usr/lib/mull-ir-frontend-18 -g -grecord-command-line -fsanitize=address,undefined"
```

Before:

```
[info] Using config uriparser/mull.yml
[warning] Could not find dynamic library: libstdc++.so.6
[warning] Could not find dynamic library: libm.so.6
[warning] Could not find dynamic library: libresolv.so.2
[warning] Could not find dynamic library: libgcc_s.so.1
[warning] Could not find dynamic library: libc.so.6
[info] Warm up run (threads: 1)
       [################################] 1/1. Finished in 45ms
[info] Filter mutants (threads: 1)
       [################################] 1/1. Finished in 0ms
[info] Baseline run (threads: 1)
       [################################] 1/1. Finished in 41ms
[info] Running mutants (threads: 12)
       [################################] 33/33. Finished in 833ms
[info] Survived mutants (7/33):
uriparser/src/UriMemory.c:191:11: warning: Survived: Replaced > with >= [cxx_gt_to_ge]
	if (size > ((size_t)-1) - extraBytes) {
	         ^
uriparser/src/UriMemory.c:209:24: warning: Survived: Replaced + with - [cxx_add_to_sub]
	return (char *)buffer + extraBytes;
	                      ^
uriparser/src/UriParseBase.c:57:33: warning: Survived: Replaced + with - [cxx_add_to_sub]
		output[1] = 16 * hexDigits[0] + hexDigits[1];
		                              ^
uriparser/src/UriParseBase.c:63:33: warning: Survived: Replaced + with - [cxx_add_to_sub]
		output[1] = 16 * hexDigits[1] + hexDigits[2];
		                              ^
uriparser/src/UriParseBase.c:83:25: warning: Survived: Replaced + with - [cxx_add_to_sub]
		return 10 * digits[0] + digits[1];
		                      ^
uriparser/src/UriParseBase.c:87:26: warning: Survived: Replaced + with - [cxx_add_to_sub]
		return 100 * digits[0] + 10 * digits[1] + digits[2];
		                       ^
uriparser/src/UriParseBase.c:87:43: warning: Survived: Replaced + with - [cxx_add_to_sub]
		return 100 * digits[0] + 10 * digits[1] + digits[2];
		                                        ^
[info] Mutation score: 78%
[info] Total execution time: 921ms
[info] Surviving mutants: 7
```

After:

```
[info] Using config uriparser/mull.yml
[warning] Could not find dynamic library: libstdc++.so.6
[warning] Could not find dynamic library: libm.so.6
[warning] Could not find dynamic library: libresolv.so.2
[warning] Could not find dynamic library: libgcc_s.so.1
[warning] Could not find dynamic library: libc.so.6
[info] Warm up run (threads: 1)
       [################################] 1/1. Finished in 46ms
[info] Filter mutants (threads: 1)
       [################################] 1/1. Finished in 0ms
[info] Baseline run (threads: 1)
       [################################] 1/1. Finished in 42ms
[info] Running mutants (threads: 12)
       [################################] 33/33. Finished in 767ms
[info] Survived mutants (2/33):
uriparser/src/UriMemory.c:191:11: warning: Survived: Replaced > with >= [cxx_gt_to_ge]
	if (size > ((size_t)-1) - extraBytes) {
	         ^
uriparser/src/UriMemory.c:209:24: warning: Survived: Replaced + with - [cxx_add_to_sub]
	return (char *)buffer + extraBytes;
	                      ^
[info] Mutation score: 93%
[info] Total execution time: 857ms
[info] Surviving mutants: 2
```

### CODE REVIEW

**Comment by TimWolla** [comment]
> I have not used Mull before

Neither did I. Mutation testing is something I've known from PHP, but not in C. After the discussion about the `if (relative)` in the other PR, I wondered if there is something for C and came across Mull which seemed to be the most reliable one. Unfortunately the number of mutators (https://mull.readthedocs.io/en/0.26.0/SupportedMutations.html) is quite limited. An “always take branch” and “never take branch” mutator would be helpful.

**Review by hartwork — CHANGES_REQUESTED** [review]
@TimWolla I had a chance at a closer look now:

- Please see the two comments on details below, and
- could you `git am` patch [0001-Add-two-more-Mull-annotations.patch](https://github.com/user-attachments/files/22730995/0001-Add-two-more-Mull-annotations.patch) for me on top to address the two remaining cases?

Thanks in advance! :pray:

File: src/UriMemory.c [hunk 1]
```diff
@@ -237,7 +237,7 @@ static void * uriDecorateRealloc(UriMemoryManager * memory,
 	prevSize = *((size_t *)((char *)ptr - sizeof(size_t) - URI_MALLOC_PADDING));
 
 	/* Anything to do? */
-	if (size <= prevSize) {
+	if (size <= prevSize) { /* mull-ignore cxx_le_to_lt */
```
[the remark this answers is not in the document]
Comment (TimWolla) [reply src/UriMemory.c:240]: I had some troubles with getting it to work with another placement. I'll try again.

[the remark this answers is not in the document]
Comment (hartwork) [reply src/UriMemory.c:240]: @TimWolla the colon version does not seem the work as documented. My guess would be, you had a colon in there earlier?

**Comment by hartwork** [comment]
For anyone else running into this pull request, here's the `Dockerfile` I used while playing with Mull locally:

```Dockerfile
# Copyright (C) 2025 Sebastian Pipping <sebastian@pipping.org>
# Licensed under Zero-Clause BSD ("0BSD")

ARG OS_VERSION=24.04
ARG LLVM_VERSION=18

FROM ubuntu:${OS_VERSION}

ARG LLVM_VERSION

RUN apt-get update && apt-get install --no-install-recommends --yes curl ca-certificates
RUN curl -1sLf 'https://dl.cloudsmith.io/public/mull-project/mull-stable/setup.deb.sh' | bash
RUN apt-get update && apt-get install --no-install-recommends --yes \
        clang-${LLVM_VERSION} \
        cmake \
        libclang-rt-${LLVM_VERSION}-dev \
        libgtest-dev \
        make \
        mull-${LLVM_VERSION}
RUN mull-runner-${LLVM_VERSION} --version

ADD cmake/*.* cmake/
ADD include/uriparser/*.h include/uriparser/
ADD src/*.c src/
ADD src/*.h src/
ADD src/*.h.in src/
ADD test/*.cpp test/

ADD CMakeLists.txt ./
ADD liburiparser.pc.in ./
ADD mull.yml ./

SHELL ["bash", "-c"]
RUN cmake \
        -DCMAKE_C_COMPILER=clang-${LLVM_VERSION} \
        -DCMAKE_CXX_COMPILER=clang++-${LLVM_VERSION} \
        -DCMAKE_{C,CXX}_FLAGS="-O0 -fpass-plugin=/usr/lib/mull-ir-frontend-${LLVM_VERSION} -g -grecord-command-line -fsanitize=address,undefined" \
        -DURIPARSER_BUILD_DOCS=OFF \
        -DURIPARSER_BUILD_TOOLS=OFF \
        -DURIPARSER_WARNINGS_AS_ERRORS=ON \
        -S . -B build
RUN make -j$(nproc) -C build
RUN mull-runner-${LLVM_VERSION} build/testrunner
```

**Comment by TimWolla** [comment]
> could you `git am` patch [0001-Add-two-more-Mull-annotations.patch](https://github.com/user-attachments/files/22730995/0001-Add-two-more-Mull-annotations.patch) for me on top to address the two remaining cases?

Unfortunately that 404s, but I just pushed my fixed-up commits. Feel free to push into my branch to include the patches.

FWIW: If that is to suppress the last two mutants, those looked legit to me, but it wasn't quite clear to me how I could kill them with tests, so I didn't suppress them.

File: mull.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+excludePaths:
+  - test/*
```
[the remark this answers is not in the document]
Comment (TimWolla) [reply mull.yml]: Duh. You are right. Should've properly read the docs :smile: That also explains why `test/*.cpp` what I initially tried didn't work.

Nevertheless this option seems to be particularly finicky, because `- ^test/` to make it a little more specific doesn't work either.

**Review by hartwork — APPROVED** [review]
@TimWolla thank you!

**Comment by hartwork** [comment]
@TimWolla PS: I started working on a test for the `size > ((size_t)-1) - extraBytes` part but it became clear the test would be rather messy without access to `URI_MALLOC_PADDING` and the function not being public API. It can be done, but the gain felt too small.

**Comment by TimWolla** [comment]
> @TimWolla okay wow, `wget https://github.com/user-attachments/files/22730995/0001-Add-two-more-Mull-annotations.patch` seems to work fine for me 🤔

It now works for me. :shrug: My only idea is that somehow GitHub didn't properly link the attachment to the comment and it then worked when you posted the same link again.

> I'm not sure if I can (or could have).

If the PR is sent from a non-organization account you generally can as a maintainer (it's an opt-out setting on PR creation). The ability is indicated by:

image

in the sidebar of a PR. I generally use `gh pr checkout 266` using [the GitHub CLI](https://cli.github.com/) to check out more complex PRs locally and then the local branch will be properly set up to push into the fork, should I want to push something. This is something I regularly do for the PHP project to merge the latest target branch to run CI and add NEWS at the same time.

> But if mutation is about holes in the test suite, then I was on the wrong mission with this.

Yes, mutation testing is to improve the test suite. Mutation testing tries to break your application by introducing “logic errors” that could have happened by accident and then checks if your tests catch these logic errors. It can be particularly useful to make sure you exercise all edge cases with regard to off-by-one errors.

In `UriParseBase.c` it exposed that we were not actually testing that IPv6 parsing works properly, because the existing test cases were “too simplistic” to properly exercise all paths. In fact even my initial version of the test case failed to kill all mutants, since I had a `210` component in the embedded IPv4 address and thus it didn't properly test that the least significant digit is properly added to the resulting address.

> PS: I started working on a test for the `size > ((size_t)-1) - extraBytes` part but it became clear the test would be rather messy without access to `URI_MALLOC_PADDING` and the function not being public API. It can be done, but the gain felt too small.

Disabling the mutator might be okay if the test would be extraordinarily complex to get rid of the noise. But it should probably be explained why in a comment. That also makes me realize that I perhaps should've added a comment for the exclusion in `uriDecorateRealloc` :thinking:
